In [2]:
!pip install transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.5/7.5 MB 15.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 268.8/268.8 kB 22.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 39.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 40.9 MB/s eta 0:00:00


In [52]:
!pip install datasets

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 519.3/519.3 kB 4.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 10.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 10.6 MB/s eta 0:00:00


In [3]:
# Use list to tokenize

In [22]:
string = 'Only those who will risk going too far can possibly find out how far one can go'

In [23]:
tokenized_str = list(string)

In [24]:
print(tokenized_str)

['O', 'n', 'l', 'y', ' ', 't', 'h', 'o', 's', 'e', ' ', 'w', 'h', 'o', ' ', 'w', 'i', 'l', 'l', ' ', 'r', 'i', 's', 'k', ' ', 'g', 'o', 'i', 'n', 'g', ' ', 't', 'o', 'o', ' ', 'f', 'a', 'r', ' ', 'c', 'a', 'n', ' ', 'p', 'o', 's', 's', 'i', 'b', 'l', 'y', ' ', 'f', 'i', 'n', 'd', ' ', 'o', 'u', 't', ' ', 'h', 'o', 'w', ' ', 'f', 'a', 'r', ' ', 'o', 'n', 'e', ' ', 'c', 'a', 'n', ' ', 'g', 'o']


In [25]:
#Numericalization

In [26]:
token2idx = {ch: idx for idx, ch in enumerate(sorted(set(tokenized_str)))}

In [27]:
print(token2idx)

{' ': 0, 'O': 1, 'a': 2, 'b': 3, 'c': 4, 'd': 5, 'e': 6, 'f': 7, 'g': 8, 'h': 9, 'i': 10, 'k': 11, 'l': 12, 'n': 13, 'o': 14, 'p': 15, 'r': 16, 's': 17, 't': 18, 'u': 19, 'w': 20, 'y': 21}


## Word Tokenization

In [28]:
tokenized_str = string.split()

In [29]:
print(tokenized_str)

['Only', 'those', 'who', 'will', 'risk', 'going', 'too', 'far', 'can', 'possibly', 'find', 'out', 'how', 'far', 'one', 'can', 'go']


In [30]:
token_word2idx = {ch:idx for idx, ch in enumerate(sorted(set(tokenized_str)))}

In [31]:
print(token_word2idx)

{'Only': 0, 'can': 1, 'far': 2, 'find': 3, 'go': 4, 'going': 5, 'how': 6, 'one': 7, 'out': 8, 'possibly': 9, 'risk': 10, 'those': 11, 'too': 12, 'who': 13, 'will': 14}


In [33]:
input_ids = [token_word2idx[token] for token in tokenized_str]

In [34]:
print(input_ids)

[0, 11, 13, 14, 10, 5, 12, 2, 1, 9, 3, 8, 6, 2, 7, 1, 4]


# Subword Tokenization


1. WordPiece - 使用基本的 Unicode 字符來建立 word 的库存, 后构建语言模型，之后把库存里面的两个word 合并生成新的word放回， 一直循环直到符合既定阈值


2. Byte Pair Encoding -  GPT 家族使用的算法。通过一个基础的word list， 对每个word 进行字符级别的排列组合，生成新的word


3. Unigram - 从一个巨大的词汇表开始，逐个删减来收缩。

## Hugging face tokenizer

In [39]:
from transformers import AutoTokenizer

In [40]:
string = "Only those who will risk going too far can possibly find out how far one can go."

In [42]:
model_name = 'distilbert-base-uncased-finetuned-sst-2-english'

In [44]:
tokenizer = AutoTokenizer.from_pretrained(model_name)

In [45]:
encoded_str = tokenizer(string, padding=True, truncation=True)

In [46]:
encoded_str

{'input_ids': [101, 2069, 2216, 2040, 2097, 3891, 2183, 2205, 2521, 2064, 4298, 2424, 2041, 2129, 2521, 2028, 2064, 2175, 1012, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

Special Token ID 下表：

Special Token	[PAD]	[UNK]	[CLS]	[SEP]	[MASK]

Special Token ID	0	100	101	102	103

In [47]:
tokens = tokenizer.convert_ids_to_tokens(encoded_str.input_ids)

In [48]:
tokens

['[CLS]',
 'only',
 'those',
 'who',
 'will',
 'risk',
 'going',
 'too',
 'far',
 'can',
 'possibly',
 'find',
 'out',
 'how',
 'far',
 'one',
 'can',
 'go',
 '.',
 '[SEP]']

In [49]:
type(encoded_str)

transformers.tokenization_utils_base.BatchEncoding

In [50]:
print(tokenizer.convert_tokens_to_string(tokens))

[CLS] only those who will risk going too far can possibly find out how far one can go. [SEP]


In [53]:
from datasets import load_dataset

In [54]:
sentiment = load_dataset("poem_sentiment")

Generating train split:   0%|          | 0/892 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/105 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/104 [00:00<?, ? examples/s]

In [56]:
def tokenize(batch):
  return tokenizer(batch["verse_text"], padding=True, truncation=True)

In [57]:
print(tokenize(sentiment['train'][:3]))

{'input_ids': [[101, 2007, 5122, 2630, 22681, 1012, 1999, 2122, 9379, 13178, 1011, 1011, 102], [101, 2009, 6223, 2061, 2146, 2004, 4212, 1996, 4542, 1010, 102, 0, 0], [101, 1998, 2008, 2003, 2339, 1010, 1996, 10459, 14045, 2154, 1010, 102, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]]}


In [58]:
sentiment_encoded = sentiment.map(tokenize, batched=True, batch_size=None)

Map:   0%|          | 0/892 [00:00<?, ? examples/s]

Map:   0%|          | 0/105 [00:00<?, ? examples/s]

Map:   0%|          | 0/104 [00:00<?, ? examples/s]

In [59]:
print(sentiment_encoded["train"].column_names)

['id', 'verse_text', 'label', 'input_ids', 'attention_mask']


In [60]:
print(sentiment_encoded["train"])

Dataset({
    features: ['id', 'verse_text', 'label', 'input_ids', 'attention_mask'],
    num_rows: 892
})


In [64]:
#sentiment_encoded["train"]["verse_text"]

In [65]:
string_arr = [
    "Only those who will risk going too far can possibly find out how far one can go.",
    "Baby shark, doo doo doo doo doo doo, Baby shark!"
]

In [66]:
inputs = tokenizer(string_arr, padding=True, truncation=True, return_tensors='pt')

In [67]:
print(inputs)

{'input_ids': tensor([[  101,  2069,  2216,  2040,  2097,  3891,  2183,  2205,  2521,  2064,
          4298,  2424,  2041,  2129,  2521,  2028,  2064,  2175,  1012,   102],
        [  101,  3336, 11420,  1010, 20160, 20160, 20160, 20160, 20160, 20160,
          1010,  3336, 11420,   999,   102,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0]])}


In [71]:
type(inputs['input_ids'])

torch.Tensor

In [72]:
from transformers.tools.text_classification import AutoModelForSequenceClassification
model = AutoModelForSequenceClassification.from_pretrained(model_name)

In [73]:
outputs = model(**inputs)

In [75]:
outputs.logits

tensor([[ 0.9738, -0.8098],
        [-1.1881,  1.2323]], grad_fn=<AddmmBackward0>)

In [77]:
import torch
predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)

In [78]:
predictions

tensor([[0.8561, 0.1439],
        [0.0816, 0.9184]], grad_fn=<SoftmaxBackward0>)

In [79]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}

In [ ]:
# It means:
# The 1st sentence: 85.6% negative
# The 2nd sentence: 91.8% positive